rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java

```diff
@@ -1044,68 +1093,72 @@ private void processReleaseRequests() {
-                // For a short time period, there could be items in the release request queue that don't have the rskTxHash
-                // (these are releases created right before the consensus rule activation, that weren't processed before its activation)
-                // We shouldn't generate the event for those releases
-                if (rskTxHash != null) {
-                    // Log the Release request
-                    eventLogger.logReleaseBtcRequested(rskTxHash.getBytes(), generatedTransaction, releaseRequest.getAmount());
-                }
-            } else {
-                releaseTransactionSet.add(generatedTransaction, rskExecutionBlock.getNumber());
-            }
+            addPegoutTxToReleaseTransactionSet(generatedTransaction, releaseTransactionSet, releaseRequest);
 
             // Mark UTXOs as spent
+            List<UTXO> selectedUTXOs = result.get().getSelectedUTXOs();
             availableUTXOs.removeAll(selectedUTXOs);
 
             // TODO: (Ariel Mendelzon, 07/12/2017)
             // TODO: Balance adjustment assumes that change output is output with index 1.
             // TODO: This will change if we implement multiple releases per BTC tx, so
             // TODO: it would eventually need to be fixed.
             // Adjust balances in edge cases
-            adjustBalancesIfChangeOutputWasDust(generatedTransaction, releaseRequest.getAmount());
+            adjustBalancesIfChangeOutputWasDust(generatedTransaction, releaseRequest.getAmount(), wallet);
 
             return true;
         });
     }
 
+    private void processPegoutsInBatch(
+        ReleaseRequestQueue releaseRequestQueue,
+        ReleaseTransactionBuilder txBuilder,
+        List<UTXO> availableUTXOs,
+        ReleaseTransactionSet releaseTransactionSet,
+        Wallet wallet
+    ) {
+        long currentBlockNumber = rskExecutionBlock.getNumber();
+        long nextPegoutCreationBlockNumber = getNextPegoutCreationBlockNumber();
+
+        if (currentBlockNumber >= nextPegoutCreationBlockNumber) {
+            // batch pegout transactions
+            Optional<ReleaseTransactionBuilder.BuildResult> result = txBuilder.buildBatchedPegouts(releaseRequestQueue.getEntries());
+
+            Coin totalPegoutValue = releaseRequestQueue.getEntries()
+                .stream()
+                .map(ReleaseRequestQueue.Entry::getAmount)
+                .reduce(Coin.ZERO, Coin::add);
+
+            if (!result.isPresent()) {
+                logger.warn(
+                    "Couldn't build a pegout BTC tx for {} pending requests with (total amount: {})",
+                    releaseRequestQueue.getEntries().size(),
+                    totalPegoutValue);
+                return;
+            }
+
+            BtcTransaction generatedTransaction = result.get().getBtcTx();
+            releaseTransactionSet.add(generatedTransaction, rskExecutionBlock.getNumber());
```

Comment:
Shouldn't you be calling `addPegoutTxToReleaseTransactionSet` here?

We'll probably change this in future tasks, but just keep in mind that we discussed using the tx hash of the transaction that called updateCollections where the pegouts were processed.

Reply:
**addPegoutTxToReleaseTransactionSet**  Requires ReleaseRequestQueue.Entry as a Param Which Cannot Be Provided By **processPegoutsInBatch**  Except I Pass Null